File: Problems/No.673_Number_of_Longest_Increasing_Subsequence/number_of_longest_increasing_subsequence.cpp
class Solution {
public:
    int findNumberOfLIS(vector<int>& nums) {
        
        // parameters
        std::vector<int> DP(nums.size(), 1);
        std::vector<int> LIS_cnt(nums.size(), 1);
        int MAX_LIS = 1;
        int MAX_LIS_cnt = 0;

        // start computing
        for (int i=0; i<nums.size(); i++) {

            for (int j=0; j<i; j++) 
                if (nums[j] < nums[i]) 
                    if (DP[i] < (DP[j] + 1)) {

                        DP[i] = DP[j] + 1;
                        LIS_cnt[i] = LIS_cnt[j];

                    }
                    else if ((DP[j] + 1) == DP[i]) LIS_cnt[i] += LIS_cnt[j];
            
            MAX_LIS = std::max(MAX_LIS, DP[i]);

        }

        for (int idx=0; idx<nums.size(); idx++)
            if (DP[idx] == MAX_LIS) MAX_LIS_cnt += LIS_cnt[idx];

        return MAX_LIS_cnt;

    }
};

File: Problems/No.673_Number_of_Longest_Increasing_Subsequence/number_of_longest_increasing_subsequence.py
class Solution(object):
    def findNumberOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """

        # parameters
        LIS_len = [1] * len(nums)
        LIS_cnt = [1] * len(nums)
        MAX_LIS = 1
        MAX_LIS_cnt = 0

        # start computing
        for i in range(len(nums)):
            for j in range(i):
                if (nums[j] < nums[i]):
                    if (LIS_len[i] < (LIS_len[j] + 1)):
                        LIS_len[i] = LIS_len[j] + 1
                        LIS_cnt[i] = LIS_cnt[j]
                    elif (LIS_len[i] == (LIS_len[j] + 1)):
                        LIS_cnt[i] += LIS_cnt[j]
            MAX_LIS = max(MAX_LIS, LIS_len[i])

        for [lis_len, lis_cnt] in zip(LIS_len, LIS_cnt):
            if (lis_len == MAX_LIS):
                MAX_LIS_cnt += lis_cnt
        
        return MAX_LIS_cnt
